### ISSUE TITLE
Reposition() moving photons below accretion disc

### ISSUE THREAD

**Issue by saultyevil** [sirocco-rt/sirocco#584]
Currently in most of the spherical TDE models, a fraction of photons are undergoing resonant scatters very close to the surface of the accretion disc, say around z ~ 1e5-1e6 cm. Due to scattering, photons now point in the -ve direction. At of end the while photon transport loop in trans_phot_single(), a call to reposition() is made which moves the photon a distance dfudge to avoid situations where the photon interacts with the same resonance. But reposition has now placed the photon below the accretion disc. After this, a call to walls() is used to check if the photon hit the boundary. 

However, when the code attempts to calculate the distance s (line 849 in photon2d.c),

```c
// pold is above the disk
s = (-(pold->x[2])) / (pold->lmn[2]);
```

s is negative, as both x[2] and lmn[2] are positive, triggering the error "walls: distance %g<0. Position %g %g %g". walls then returns istat = -1 and so the transport of this photon is ended.

Here is a pf file which replicates this problem after around ~35,000 photons have been transported. 

[tde_spherical_walls.pf.txt](https://github.com/agnwinds/python/files/3485854/tde_spherical_walls.pf.txt)

**Comment by saultyevil** [comment]
Sorry, but this is going to be a bit of a brain dump...

To summarise this problem.. disk photons are emitted and then interact near the very edge of the cell, i.e. say at z ~ 1e6 cm. It's possible for some photons to now point in the opposite direction, i.e. pointing in +ve z to -ve z. If the interaction was with a resonance line, then the photon is moved some push through distance dfudge, typically ~1e7 cm from my TDE models, which in some cases causes the photon to pass through the disc plane without the code being able to properly handle this situation.

For now, since this only seems to effect a small fraction of photons, the fix I've implemented in my fork is to catch this odd behaviour and tag these photons as `P_ERROR` and being lost to DFUDGE. Essentially saying that these photons have undergone some undefined behaviour.

But say if you want these photons to be reflected by the disk instead, it doesn't seem like there is an easy way to do this without re-writing or changing the order of trans_phot_single(). What I mean is that after a photon is returned from translate(), Python checks if `istat == P_HIT_STAR`, then `istat == P_HIT_DISK` and finally `istat == P_SCAT`. I'm not sure if this has to be done in this exact order or not. But, if `istat == P_SCAT` is checked first, we could cope with the situation where if a photon hits the disk after reposition() and `istat = P_HIT_DISK` is set, then when Python reaches `if (istat == P_HIT_DISK)`, the problem photons will be refelected or absorbed as normal.

Anyway, after chatting to Nick for a bit about this problem, we brainstormed a few ways with how we can deal with this better. The main thing we agreed on is that photons in reposition() should be repositioned using a physically motivated number, rather than dfudge which is geometrically motivated. For example, photons could be repositioned by some multiple of the Sobolev length scale instead. But we also both agreed that if a photon is scattering in the skin of a cell after emission, then we're likely going to encounter a plethora of other problems rather than just this one - i.e. problems associated with high optical depths.

I think this is something we should discuss more during a telecon.

**Comment by saultyevil** [comment]
A band-aid fix has been applied with PR #599 and so this issue shouldn't happen anymore.

**Comment by kslong** [comment]
We seem to be encountering a new variant of this issue the tde models.  I am reopening the issue until we can determine whether that is the case.

**Comment by saultyevil** [comment]
I presume this was re-opened due to the amount of photons in the Hit Surface spectra discussed in the call yesterday.

As far as I can tell, photons are being tagged as P_HIT_STAR as in the call to  `walls ()` [here](https://github.com/agnwinds/python/blob/82579fc6e1fba0771258c1aba77ec23868f6825a/source/trans_phot.c#L592) they are triggering this check, 

 ```c
else if (s < VERY_BIG && (test_s = ds_to_sphere (geo.rstar, p)) == VERY_BIG)
{
    /* then we hit the star somewhere in between */
    hit_star = TRUE;
}
if (hit_star == TRUE)
{
    stuff_phot (pold, p);
    move_phot (p, s);
    stuff_v (p->x, normal);
    return (p->istat = P_HIT_STAR);
}

````
Then istat = P_HIT_DISK is returned and photon transport is broken for the same reason as discussed previously (the `while (istat == P_INWIND)` condition is broken). It's curious because previously they are considered in the wind, so it's possible there is something weird happening here.

I don't think it is entirely related to reposition as it was in the original issue though. Looking at the history of some of the photons reveals that nres can be -1, meaning that reposition will not move the photon by dfudge.

I plotted the positions for one of these photons..

![Screenshot from 2019-11-20 14-43-15](https://user-images.githubusercontent.com/19627279/69248391-21554580-0ba4-11ea-93ae-91477071622e.png)

I'm not exactly sure why that if statement is triggering, but when it does, the photon is moved to the surface of the star (green diamond).

**Comment by kslong** [comment]
I think I agree, or at least my initial tests seem to show, that the P_REPOSITION_ERROR is not being triggered, but somehow the photon is still ending up with a status of HIT_STAR or HIT_DISK, so something is wrong in this section of the routine ….

**Comment by kslong** [comment]
Part of the problem related to the star at least is in walls, and in calculating whether a photon has hit the star.  walls is intended to determine both whether a photon is within the star, and whether it has hit the star along its path.  There was a reasonably recent fix which I made, and it was intended to prevent one from going a long distance which completely crossed the star.  The solution in walls works for the case where the photon direction has not changed between pold and p, but does not work in this instance, where the photon direction has been changed by a resonance interaction.  Somewhere there should be an issue about this but I could not find it.  

There are a couple of possibilities here.  One would be to make an assumption that if the old and new photon are going in different directions, that we are not trying to to look for a problem along the path.  The other would be to leave out the check of walls at the end of trans_phot_single, but a lot of this is bound up in the logic to check whether we are where we expected to be after repositioning.  

This is sufficiently important that we need to really understand what is happening and not simpy band aide the code.
